### PR TITLE
Methods branch

### PR DESCRIPTION
**Pull request by RoodzFernando**
The enumerable module sent for code review. Thank You!

### CODE REVIEW

File: enumerables.rb [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+          temp_array << item if yield(item)
+        end
+        temp_array
+      end
+    else
+      to_enum
+    end
+  end
+
+  # my_all?
```
Comment (ebukaume) [inline enumerables.rb]: These comments are not necessary. Your method definitions are just enough here.

File: enumerables.rb [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+      end
+      (0...temp_array.size).each do |i|
+        return true if temp_array[i] == true
+
+        return false
+      end
+    else
+      to_enum
+    end
+  end
```
Comment (ebukaume) [inline enumerables.rb]: In ruby, all methods that end in `?` return Boolean. Try this in your environment:

`[1, :s, 'a', {}].all?`

Is your own implementation working like ruby all? method? :wink: 

Also, `all?` optionally accepts an argument but your implementation does not. Also, account for different kinds of input types. Refer to [the docs](https://devdocs.io/ruby~2.6/enumerable)

File: enumerables.rb [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+        temp_array << yield(item)
+      end
+      (0...temp_array.size).each do |i|
+        return true if temp_array[i] == true || temp_array.nil?
+      end
+    else
+      true
+    end
+  end
+
```
Comment (ebukaume) [inline enumerables.rb]: Same as above

File: enumerables.rb [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+      my_each do |item|
+        temp_array << yield(item)
+      end
+      (0...temp_array.size).each do |i|
+        return true if temp_array[i] == false || temp_array.nil?
+      end
+    else
+      true
+    end
+  end
```
Comment (ebukaume) [inline enumerables.rb]: Same as above

File: enumerables.rb [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+  def my_count
+    if instance_of?(Array)
+      temp_array = []
+      my_each do |item|
+        temp_array << item if yield(item)
+      end
+      temp_array.size
+    end
+    return unless block_given?
+  end
```
Comment (ebukaume) [inline enumerables.rb]: What happens when block is not given? `[1, 2, :s, 'a'].count` 

Or when the user passes an argument? [1, 2, :s, 'a'].count(1)`. 

Did you consider when proc is passed? [1, 2, :s, 'a'].count(&some_procedure)`

File: enumerables.rb [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+  #    if block_given?
+  #         temp_array = []
+  #         self.my_each do |item|
+  #             temp_array << block.call(item)
+  #         end
+  #         temp_array
+  #         else
+  #         self.to_enum
+  #     end
+  # end
```
Comment (ebukaume) [inline enumerables.rb]: Remove this commented code. If you need to review a previous version of your code, use git history.

File: enumerables.rb [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+        return result
+      else
+        result = 0
+        i = param
+        (i...size).each do |n|
+          result = yield(result, self[n])
+        end
+      end
+    else
+      puts 'No block given'
```
Comment (ebukaume) [inline enumerables.rb]: No debugging information should be pushed to the repo. Please remove it. You can raise an error if need be.

File: enumerables.rb [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+      if block_given?
+        yield(arr[i])
+      else
+        result << proc.call(arr[i])
+      end
+      i += 1
+    end
+    result
+  end
+end
```
Comment (ebukaume) [inline enumerables.rb]: What happens when a proc is passed? What of when neither a block nor an argument is passed? Compare it with the ruby `#map` method.

File: enumerables.rb [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+          result = yield(result, self[index])
+        end
+        return result
+      else
+        result = 0
+        i = param
+        (i...size).each do |n|
+          result = yield(result, self[n])
+        end
+      end
```
Comment (ebukaume) [inline enumerables.rb]: Did you test for when it is called on a range? Remember, the Range class also includes the Enumerable method.

Also, what should happen when a block is not given? 

`[1,2,3,4,5].my_inject(:+)` should equal `15` not `nil`

**Review by ebukaume — CHANGES_REQUESTED** [review]
Hi @RoodzFernando :wave: 

Nice job so far. However, most of your methods are not behaving the way they should. You may need to look at the docs to have a solid grasp of the expected behaviours of ruby enumerable methods.

_See below for more_

**Kindly submit another code review after fixing the issues.**

Happy Coding :v: 
**Ebuka Umeokonkwo**
 [Github](https://github.com/ebukaume) | [Linkedin](https://www.linkedin.com/in/ebukaume) | [Slack](https://app.slack.com/team/UFDKXE68L) | [Twitter](https://twitter.com/ebukaume)

**Comment by RoodzFernando** [comment]
I have all the bugs fixed!

File: enumerables.rb [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+      if pattern.class == Regexp
+        my_each { |x| return true unless pattern.match?(x) }
+      elsif pattern.class == Class
+        my_each { |x| return true unless x.is_a? pattern }
+      else
+        my_each { |x| return true unless x == pattern }
+      end
+    end
+    false
+  end
```
Comment (stratospherique) [inline enumerables.rb]: * [ ] my_any? should be able to receive classes, literals, regex exp, etc as **arguments** :no_entry_sign: 

* [ ] my_any? returns true if the block ever returns a value other than false or nil
  ```ruby
    array = [1,2,5,4,7]
    true_block = proc {|num| num < 10 }
    p array.my_any?(&true_block) == array.any?(&true_block) # false
  ```

File: enumerables.rb [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+  def my_none?
+    if block_given?
+      !my_any? { |x| yield x }
+    elsif pattern
+      !my_any?(pattern)
+    else
+      my_each { |x| return false if x }
+      true
+    end
+  end
```
Comment (stratospherique) [inline enumerables.rb]: * [ ] my_none? should be able to receive classes, literals, regex exp, etc as **arguments** :no_entry_sign: 

* [ ] my_none? returns true if the block never returns true for all elements
  ```ruby
    array = [1,2,5,4,7]
    true_block = proc {|num| num > 10 }
    p array.my_none?(&true_block) == array.none?(&true_block) # false
  ```

File: enumerables.rb [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+    count = 0
+    if block_given?
+      my_each { |x| count += 1 if yield x }
+    elsif arg
+      my_each { |x| count += 1 if x == arg }
+    else
+      my_each { count += 1 }
+    end
+    count
+  end
```
Comment (stratospherique) [inline enumerables.rb]: * [x] my_count returns the number of items in enum through enumeration
  ```ruby
    array = [1,2,5,4,7]
    p array.my_count == array.count # returns this error
      #
      # NameError:
      #  undefined local variable or method `arg' for #<Array:0x000055b627046ab8>
      #  Did you mean?  _args
      # ./enumerables.rb:91:in `my_count'
  ```    

* [x] my_count counts the number of items in enum that are equal to item if an argument is given 
     ```ruby
    array = [1,2,5,4,7,1,8,7,1]
    p array.my_count(1) == array.count(1) # returns this error
      #
      # NameError:
      #  undefined local variable or method `arg' for #<Array:0x000055b627046ab8>
      #  Did you mean?  _args
      # ./enumerables.rb:91:in `my_count'
  ```

File: enumerables.rb [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+  def multiply_els(arr)
+    arr.my_inject { |n, p| n * p }
+  end
+
+  def my_map
+    result = []
+    (1..size).each { |n| result << yield(n) } if self.class == Range
+    (1..size).each { |n| result << yield(n) } if self.class != Range
+    result
+  end
```
Comment (stratospherique) [inline enumerables.rb]: * [x] my_map returns an Enumerator if no block is given
   ```ruby
    array = [1,5,8,7,9,5,5,55,5]
    p array.my_map #error
    LocalJumpError:
        no block given (yield)
      # ./Enumerable_Methods.rb:89:in `block in my_map'
      # ./Enumerable_Methods.rb:12:in `my_each'
      # ./Enumerable_Methods.rb:88:in `my_map'
  ```

* [x] my_map returns a new array with the results of running block once for every element in enum.
  ```ruby
    array = [1,2,5,4,7]
    block = proc { |num| num < (1 + 9) / 2 }
    p array.my_map?(&block) == array.map?(&block) # false
  ```

File: enumerables.rb [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+      return my_inject_sym(initial) if initial.class == Symbol
+      return my_inject_sym(sym, initial) if sym
+
+      raise 'No block nor symbol given'
+    end
+    return self[1..length].my_inject(self[0], &block) unless initial
+
+    my_each { |x| initial = block.call(initial, x) }
+    initial
+  end
```
Comment (stratospherique) [inline enumerables.rb]: * [ ] my_inject combines all elements of enum by applying a binary operation, specified by a block
  ```ruby
    range = Range.new(5, 50)
    p range.my_inject(4) { |prod, n| prod * n } # returns this error
    #NoMethodError:
    #    undefined method `[]' for 5..50:Range
      # ./enumerables.rb:11:in `block in my_each'
      # ./enumerables.rb:10:in `each'
      # ./enumerables.rb:10:in `my_each'
      # ./enumerables.rb:108:in `my_inject'
  ```

* [ ] my_inject when a symbol is specified combines each element of the collection by applying the symbol as a named method
  ```ruby
    range = Range.new(5, 50)
    p range.my_inject(2, :*) # returns this error
    #NoMethodError:
    #    undefined method `[]' for 5..50:Range
      # ./enumerables.rb:11:in `block in my_each'
      # ./enumerables.rb:10:in `each'
      # ./enumerables.rb:10:in `my_each'
      # ./enumerables.rb:115:in `my_inject_sym'
      # ./enumerables.rb:102:in `my_inject'
  ```

**Review by stratospherique — CHANGES_REQUESTED** [review]
### Great Progress @RoodzFernando :clap: However, after the last commited changes, some errors have emerged for some edgy test cases. check :arrow_down: 

- **[REQUIRED]** check the code comments below, use the notes and examples to refactor your code :gear: 

### That's it :relieved: When finished, kindly resubmit another code review request :pray: 

Happy coding :v: 
**_Ahmed Mahfoudh_**

File: enumerables.rb [hunk 15]
```diff
@@ -0,0 +1,124 @@
...
+      result = []
+      (1..size).each { |n| result << yield(n) } if self.class == Range
+      (0...size).each { |n| result << yield(self[n].to_i) } if self.class != Range
+      (0...size).each { |n| result << yield(self[n].to_i) } if self.class == Proc
+      result
+    else
+      to_enum
+    end
+  end
+end
```
Comment (Forison) [inline enumerables.rb:137]: ` array = [1,2,5,4,7] `
  `block = proc { |num| num < (1 + 9) / 2 }`
  `p array.my_map?(&block) =  array.map?(&block)` # false`

File: enumerables.rb [hunk 16]
```diff
@@ -0,0 +1,124 @@
...
+    count
+  end
+
+  def my_inject(*_item)
+    acc = self[0]
+    self[1..length].each do |x|
+      acc = yield(acc, x)
+    end
+    acc
+  end
```
Comment (Forison) [inline enumerables.rb:120]: `range = Range.new(5, 50)`
  ` p range.to_a.my_inject(4) { |prod, n| prod * n }`   #still returns this error
   compare to the real ```inject```
  `p range.inject(4) { |prod, n| prod * n }`

Comment (Forison) [inline enumerables.rb:120]: ` range = Range.new(5, 50) `
 ` p range.my_inject(2, :*) ` # returns this error
compare it to the result you get using the real ```inject```
 ` p range.inject(2, :*) `

**Review by Forison — CHANGES_REQUESTED** [review]
# Hi @RoodzFernando,
Good job on implementing the changes some of the changes suggested in the previous review.
Here are a few flaws you have to address to have the project approved💪.
See below⬇️
## That is it, Kindly do these changes and submit the project for review
#### Happy coding😇
#### @Forison

**Comment by RoodzFernando** [comment]
p array.my_map?(&block) = array.map?(&block) # false` (This enumerable works for me maybe it is because the original map does not take the question at his end.) (https://devdocs.io/ruby~2.6/enumerable#method-i-map) check the documentation please fro more information. Thank you!

**Review by whiz25 — CHANGES_REQUESTED** [review]
## Good job :+1: 
Nice work implementing the suggested changes. I tested your methods and they cover some edge cases which is good. However, we can't merge this one yet. So check my comments for changes to be made and resend for another review.

### Martin Afani

File: enumerables.rb [hunk 17]
```diff
@@ -0,0 +1,139 @@
...
+      if pattern.class == Regexp
+        my_each { |x| return false unless pattern.match?(x) }
+      elsif pattern.class == Class
+        my_each { |x| return false unless x.is_a? pattern }
+      else
+        my_each { |x| return false unless x == pattern }
+      end
+    end
+    true
+  end
```
Comment (whiz25) [inline enumerables.rb]: - [ ] when a class is passed as an argument returns true if all of the collection is a member of such class
- [ ] when a Regex is passed as an argument returns true if all of the collection matches the Regex
- [ ] when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern

File: enumerables.rb [hunk 18]
```diff
@@ -0,0 +1,139 @@
...
+      if pattern.class == Regexp
+        my_each { |x| return true unless pattern.match?(x) }
+      elsif pattern.class == Class
+        my_each { |x| return true unless x.is_a? pattern }
+      else
+        my_each { |x| return true unless x == pattern }
+      end
+    end
+    false
+  end
```
Comment (whiz25) [inline enumerables.rb]: - [ ] returns true if the block ever returns a value other than false or nil
- [ ] when no block or argument is given returns true if at least one of the collection is not false or nil
- [ ] when a class is passed as an argument returns true if at least one of the collection is a member of such class
- [ ] when a Regex is passed as an argument returns false if none of the collection matches the Regex
- [ ] when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern

File: enumerables.rb [hunk 19]
```diff
@@ -0,0 +1,139 @@
...
+  def my_none?
+    if block_given?
+      !my_any? { |x| yield x }
+    elsif pattern
+      !my_any?(pattern)
+    else
+      my_each { |x| return false if x }
+      true
+    end
+  end
```
Comment (whiz25) [inline enumerables.rb]: - [ ] returns true if the block never returns true for all elements
- [ ] when no block or argument is given returns true only if none of the collection members is true
- [ ] when a class is passed as an argument returns true if none of the collection is a member of such class
- [ ] when a Regex is passed as an argument returns true only if none of the collection matches the Regex
- [ ] when a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern

File: enumerables.rb [hunk 20]
```diff
@@ -0,0 +1,139 @@
...
+      initial = nil
+    end
+    total = initial
+    total = instance.shift if initial.nil?
+    instance.each do |n|
+      total = operation_hash[sym].call(total, n) if initial.nil?
+      total = yield(total, n) if block_given?
+    end
+    total
+  end
```
Comment (whiz25) [inline enumerables.rb:120]: - [ ] combines all elements of enum by applying a binary operation, specified by a block
- [ ] when a symbol is specified combines each element of the collection by applying the symbol as a named method

**Review by imhta — CHANGES_REQUESTED** [review]
# Changes Required 🛑 

### Hello @RoodzFernando, 👋  Still you didn't fix the previous code review comments. and your project is failing in the same tests like before. So kindly fix and then make another code review request.

**Happy coding,**
 Pon Muthu Selvam N
_If you have any queries feel free ping me in slack_

**Review by whiz25 — APPROVED** [review]
## Hey @RoodzFernando 
You haven't made any changes as far the previous reviews are concerned. The same edge cases are still failing. But I am giving you the :+1: because this has been a long review for you and therefore the max number of review cycle has been reached.
#### Please try to make the suggested changes for your learning.

### Martin Afani

**Pull request merged**